### PR TITLE
test_osk: add z-based runs

### PR DESCRIPTION
**Pull request by Saviq**

### CODE REVIEW

File: mir-ci/mir_ci/tests/robot/platforms/zapper/ZapperHid.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        Walk the virtual pointer to an absolute position within the output, maximum `step_distance`
+        at a time, with `delay` seconds in between.
+        """
+        assert isinstance(x, int) and isinstance(
+            y, int
+        ), "Coordinates must be integers"
+
+        resolution = _get_resolution()
+        assert 0 <= x <= resolution[0], "X coordinate outside of screen"
+        assert 0 <= y <= resolution[1], "Y coordinate outside of screen"
```
Comment (hbatagelo) [inline mir-ci/mir_ci/tests/robot/platforms/zapper/ZapperHid.py:134]: I wonder if it should be a right-open interval instead. In a 100x100 output, I would expect the in-screen coordinates to be in the range of 0 to 99. OTOH, the current conversions from proportional to absolute coordinates assume that 0..1 is converted to 0..100.

**Review by hbatagelo — CHANGES_REQUESTED** [review]

File: mir-ci/mir_ci/tests/robot/platforms/wayland/WaylandHid.py [hunk 2]
```diff
@@ -79,6 +79,12 @@ async def release_pointer_button(self, button: str) -> None:
...
         self.button(Button[button], False)
 
+    @keyword
+    async def click_pointer_button(self, button: str) -> None:
+        await self.connect()
+        self.button(Button[button], True)
+        self.button(Button[button], False)
+
     @keyword
     async def release_buttons(self) -> None:
```
[the remark this answers is not in the document]
Comment (Saviq) [reply mir-ci/mir_ci/tests/robot/platforms/wayland/WaylandHid.py:89]: Good catch. Fixed.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -1,4 +1,9 @@
-exclude: ^bin|mir-ci/mir_ci/wayland/protocols
+exclude: |
+  (?x)^(
+    bin/.*|
+    mir-ci/mir_ci/wayland/protocols/.*|
+    mir-ci/mir_ci/tests/robot/platforms/zapper/ZapperHid.py
```
Comment (mattkae) [inline .pre-commit-config.yaml:5]: Why is `ZapperHid.py` excluded here?

**Review by mattkae — CHANGES_REQUESTED** [review]
Minor things, mostly questions! I don't understand the `.resource` files well enough to have any significant feedback ther

File: .pre-commit-config.yaml (hunk 3 above)
Comment (Saviq) [reply .pre-commit-config.yaml:5]: It's only temporarily here, going into zapper proper soon, didn't want to deal with the different formatting between here and there.

File: mir-ci/mir_ci/tests/test_osk.py [hunk 4]
```diff
@@ -41,31 +42,81 @@ def collect_assets(platform: str, resources: Collection[str], suite: str, varian
...
+    "server",
+    servers.servers(
+        servers.ServerCap.INPUT_METHOD,
+        {
+            servers.confined_shell: {
+                "marks": (pytest.mark.xfail(reason="OSK not activating", raises=ProgramError, strict=True),)
+            },
+            servers.miriway: {
+                "marks": (pytest.mark.xfail(reason="OSK not activating", raises=ProgramError, strict=True),)
+            },
```
[the remark this answers is not in the document]
Comment (Saviq) [reply mir-ci/mir_ci/tests/test_osk.py:79]: Not yet, I still need to find a reproducer better than "run this test from mir-ci".

File: mir-ci/mir_ci/tests/test_osk.py [hunk 5]
```diff
@@ -41,31 +42,81 @@ def collect_assets(platform: str, resources: Collection[str], suite: str, varian
...
             server,
             add_extensions=extensions,
         )
-
-        robot = server_instance.program(App(("robot", "-d", tmp_path, tmp_path)))
-
-        assets = collect_assets("wayland", ("osk",), "osk")
-
-        tuple((tmp_path / k).symlink_to(v) for k, v in assets.items())
+        assets = collect_assets(platform, ("kvm", "osk"), "osk")
```
[the remark this answers is not in the document]
Comment (Saviq) [reply mir-ci/mir_ci/tests/test_osk.py:92]: Not yet, but as we move more tests into using this, we'll end up with variants per display server, for example.

File: mir-ci/mir_ci/tests/test_osk.py [hunk 6]
```diff
@@ -57,21 +83,40 @@ def collect_assets(platform: str, resources: Collection[str], suite: str, varian
...
+                tuple((tmp_path / k).symlink_to(v) for k, v in assets.items())
+                robot = server_instance.program(App(("robot", "-d", tmp_path, tmp_path)))
+                async with robot:
+                    await robot.wait(120)
 
-        assets = collect_assets("wayland", ("osk",), "osk")
+            elif platform == "zapper":
+                proxy = import_module("checkbox_support.scripts.zapper_proxy")
+                _robot = b""
+                _assets = {}
```
[the remark this answers is not in the document]
Comment (Saviq) [reply mir-ci/mir_ci/tests/test_osk.py:104]: Just going in circles about local / private variables (didn't want to shadow `assets` from above), and `_robot` was just in the line of fire...

[the remark this answers is not in the document]
Comment (Saviq) [reply mir-ci/mir_ci/tests/test_osk.py]: Fixed.

**Review by hbatagelo — APPROVED** [review]

**Review by mattkae — APPROVED** [review]
Makes sense to me!

**Pull request merged**